packages/core/src/modules/connections/DidExchangeProtocol.ts

```diff
@@ -0,0 +1,462 @@
+      didDocument = JsonTransformer.fromJSON(message.didDoc.getDataAsJson(), DidDocument)
+    } else {
+      didDocument = peerDid.didDocument
+    }
+
+    const didRecord = new DidRecord({
+      id: message.did,
+      role: DidDocumentRole.Received,
+      // It is important to take the did document from the PeerDid class
+      // as it will have the id property
+      didDocument,
+      tags: {
+        // We need to save the recipientKeys, so we can find the associated did
+        // of a key when we receive a message from another connection.
+        recipientKeys: didDocument.recipientKeys,
+      },
+    })
+
+    await this.didRepository.save(didRecord)
+
+    connectionRecord.theirDid = message.did
+    connectionRecord.theirLabel = message.label
+    connectionRecord.threadId = message.id
+
+    await this.updateState(DidExchangeRequestMessage.type, connectionRecord)
+    this.logger.debug(`Process message ${DidExchangeRequestMessage.type} end`, connectionRecord)
+    return connectionRecord
+  }
+
+  public async createResponse(
+    connectionRecord: ConnectionRecord,
+    routing?: Routing
+  ): Promise<DidExchangeResponseMessage> {
+    this.logger.debug(`Create message ${DidExchangeResponseMessage.type} start`, connectionRecord)
+    DidExchangeStateMachine.assertCreateMessageState(DidExchangeResponseMessage.type, connectionRecord)
+
+    // They will then either update the provisional service information to rotate the key, or provision a new DID entirely.
+    // The choice here will depend on the nature of the DID used in the invitation.
+
+    // if reuse did from invitation then do ...
+    // otherwise create new did and didDoc
+    const { did, threadId } = connectionRecord
+
+    if (!threadId) {
+      throw new AriesFrameworkError('Missing threadId on connection record.')
+    }
+
+    // The responder will then craft an exchange response using the newly updated or provisioned information.
+
+    // Sign message attachment
+    // Use invitationKey by default, fall back to verkey (?)
+    const [verkey] = connectionRecord.invitation?.recipientKeys || []
+
+    if (!verkey) {
+      throw new AriesFrameworkError('Connection invitation does not contain recipient key.')
+    }
+
+    const peerDidRouting = routing || {
+      endpoints: connectionRecord.invitation?.serviceEndpoint ? [connectionRecord.invitation?.serviceEndpoint] : [],
+      verkey,
+      did,
+      routingKeys: connectionRecord.invitation?.routingKeys || [],
+    }
+
+    // TODO Currently, we just reuse invitation to create a peer did, we should also add option to create new keys
+    const peerDid = await this.createPeerDidDoc(peerDidRouting)
+    connectionRecord.did = peerDid.did
+
+    const message = new DidExchangeResponseMessage({ did: peerDid.did, threadId })
+
+    // TODO
+    // As I understood, a numAlgo 0 doesn't include service inside encoded peer did therefore we should also create a did doc attachment for it
```

Comment:
numAlgo 0 shoudn't be used for DIDExchange as it can't contain a service if using static peer dids (which we do).

Reply:
Yes, I updated it to support only numalgo 1 for now.